Formatter docs (#837)
deepchecks/deepchecks: docs/source/user-guide/formatter_objects.rst

.. formatters:

.. currentmodule:: deepchecks.vision.utils

=============================================

> The Image, Labels, and Prediction formatters

Reviewer: ```suggestion
The Image, Labels, and Prediction Formatters
```

=============================================

The formatter objects in the deepchecks.vision module were designed to enable access to the data while making the formatting easy and maintainable.
They are responsible for converting the data (whether these are images, labels, or predictions) to a format deepchecks can work with.

> For checks that works on the raw images, see :ref:`Image Formatters <formatters__image_formatter>`.

Reviewer: ```suggestion
For checks that need the image's data, see :ref:`Image Formatters <formatters__image_formatter>`.
```

> For classification tasks, see :ref:`Classification Formatters <formatters__classification_formatter>`.

Reviewer: ```suggestion
For prediction and label formats for **classification tasks**, see :ref:`Classification Formatters <formatters__classification_formatter>`.
```

> For object detection tasks, see :ref:`Object Detection Formatters <formatters__object_detection_formatter>`.

Reviewer: ```suggestion
For prediction and label formats for **object detection** tasks, see :ref:`Object Detection Formatters <formatters__object_detection_formatter>`.
```

Why is this important?
========================
There is no standard when it comes to data formats in computer vision. Each model can have its own supported data format for the output predictions, and each new benchmark dataset can propose a new labeling format. Even the images formats can be different from each other (PIL, OpenCV, etc.).

In order to run a model and data-agnostic comprehensive testing suites, the data should be in an accepted format the checks can work with.

> The formatter objects provide an easy way to do that, and are an important part of the vision module.

Reviewer: ```suggestion
The formatter objects provide a structured and repeatable way to do that, and are an important part of the vision module.
```

Types of formatters
==========================
Deepchecks requires three data formatters:

- **Image Formatter** - responsible for formatting the input images to the accepted format
- **Prediction Formatter** - responsible for formatting the output of the model to an accepted format (depends on the model type)
- **Label Formatter** - responsible for formatting the targets to an accepted format (depends on the task type)

In order to initialize a new formatter, a transformation function between the input format and the accepted format is required.

.. _formatters__image_formatter:

The Image Formatter
=====================

> The image formatter is responsible for formatting the image data outputted from the data loader to the required format for check displays.

Reviewer: ```suggestion
The image formatter is responsible for transforming the image data that is received as an output from the data loader to the required format for check's display (e.g. unnormalize the image).
```

In order to create an image formatter, you need to initialize an instance of the :class:`image_formatters.ImageFormatter` class,
passing the transformation function.

Task types formats
=====================
Each CV task type (e.g. classification, object detection, instance segmentation, etc.) has its own data formats.
For example, in classification tasks a common format for the label is simply the label_id, and for the prediction is a probability vector with the probability for each class. In object detection, there are various formats like the COCO format, PASCAL VOC XML format, YOLO format, and more...

Deepchecks accepts a specific data format for each task type and for some types, provides utility functions to convert between common formats to the accepted one.

.. _formatters__classification_formatter:

Classification format
-------------------------

Accepted format
~~~~~~~~~~~~~~~~~~~~~

> In classification tasks, it is required to implement the following formatters:

Reviewer: If we work according to the COCO formats, why not write this here? (and also in the Object detection one)

> In classification tasks, it is required to implement the following formatters:

Reviewer: ```suggestion
For classification tasks, it is required to implement the following formatters:
```

- **Label**  - should implement an instance of :class:`classification_formatters.ClassificationLabelFormatter`.
- **Predictions** - should implement an instance of :class:`classification_formatters.ClassificationPredictionFormatter`.

.. _formatters__object_detection_formatter:

Object detection format
-------------------------

Accepted Format
~~~~~~~~~~~~~~~~

> In object detection tasks, it is required to implement the following formatters:

Reviewer: ```suggestion
For object detection tasks, it is required to implement the following formatters:
```

- **Label** - should implement an instance of :class:`detection_formatters.DetectionLabelFormatter`.

> - **Detections** - should implement an instance of :class:`detection_formatters.DetectionPredictionFormatter`.

Reviewer: Shouldn't this be called also **Predictions**?
or maybe **Detection Predictions**

Converting from common formats
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> TBD (Not implemented yet)

Reviewer: ```suggestion
In order to support more use-cases with less lines of code, we have implemented built-in formatters for many of the popular formats. For object detection, these will include the following supported formats:
            - 'xyxy' - x1, y1, x2, y2 represent the upper left and lower right corners of the bounding box.
            - 'xywh' - x1, y1, w, h represent the upper left corner of the bounding box and its width and height.
            - 'cxcywh' - cx, cy, w, h represent the center of the bounding box and its width and height.
            - 'xyxyn' - x1, y1, x2, y2, n represent the upper left and lower right corners of the bounding box,
                        normalized to the image dimensions.
            - 'xywhn' - x1, y1, w, h, represent the upper left corner of the bounding box and its width and height,
                        normalized by the image dimensions.
            - 'cxcywhn' - (YOLO format) x, y, w, h, represent the center of the bounding box and its width and height,
                          normalized by the image dimensions.
If your use case is in one of the supported formats please reach out to us and we can send you the required function (pre-release) and save you the implementation, since this is still in development. Alternatively, if you think your format is common but is not listed here, please reach out and we will consider adding it (Of course, issues on github and contributions after we release are also welcome).
```

Reviewer: btw I'm considering delaying this task until we actually see that there are people that request it, WDYT?

Author: I agree. we shouldn't have it

The Image Formatter
=====================

> The image formatter is responsible for transforming the image data that is received as an output from the data loader to the required format for check's display (e.g. unnormalize the image).

Reviewer: ```suggestion
The image formatter is responsible for transforming the image data that is received as an output from the data loader to the required format for the check's display (e.g. unnormalize the image).
```
